Add windows plugin texture support

jnschulze opened the pull request:
## Description

This PR adds OpenGL texture support for plugins on the Windows platform.

Once this has landed flutter/flutter#61098 needs to be merged as well.

## Related Issues

- flutter/flutter#38601
- #9822
  

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [contributor guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [C++, Objective-C, Java style guides] for the engine.
- [x] I read the [tree hygiene] wiki page, which explains my responsibilities.
- [ ] I updated/added relevant documentation.
- [x] All existing and new tests are passing.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[contributor guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[tree hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

jnschulze commented:
@stuartmorgan 
@chinmaygarde 

Could you review this PR please?

cloudwebrtc commented:
@jnschulze I tested the [flutter-webrtc](https://github.com/cloudwebrtc/flutter-webrtc/pull/290) plugin for windows and it works. You did cool things.

jnschulze commented:
@cloudwebrtc 
I‘m glad it‘s working. But since this is based on your GLFW texture support PR (#9822) the credit goes to you.

clarkezone commented:
#18878 has landed now.  Please update to reflect that, thx.

jnschulze commented:
@clarkezone I've just updated the PR. Please have a look.

jnschulze commented:
@clarkezone I've just created [an example plugin](https://github.com/jnschulze/flutter-playground/blob/master/windows_texture_test) that renders EBU color bars.

![image](https://user-images.githubusercontent.com/720469/87244983-cc6f8380-c441-11ea-9ac3-b1db50f0a01c.png)

stuartmorgan commented on shell/platform/common/cpp/public/flutter_plugin_registrar.h:
```diff
@@ -26,6 +27,10 @@ typedef void (*FlutterDesktopOnRegistrarDestroyed)(
 
+// Returns the texture registrar associated with this registrar.
+FLUTTER_EXPORT FlutterDesktopTextureRegistrarRef
+FlutterDesktopGetTextureRegistrar(FlutterDesktopPluginRegistrarRef registrar);
```
This should be called `FlutterDesktopRegistrarGetTextureRegistrar` for consistency with the other plugin registrar methods. Although now that there are two registrars, we should really s/Registrar/PluginRegistrar/ in all these function name prefixes.

stuartmorgan commented on shell/platform/common/cpp/public/flutter_texture_registrar.h:
```diff
@@ -0,0 +1,62 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
```
Note: you'll need to update flutter/flutter [in several places](https://github.com/flutter/flutter/search?q=flutter_plugin_registrar.h&unscoped_q=flutter_plugin_registrar.h) once this lands.

stuartmorgan commented on shell/platform/common/cpp/public/flutter_texture_registrar.h:
```diff
@@ -0,0 +1,62 @@
+  size_t height;
+} PixelBuffer;
+
+// The pixel buffer copy callback definition is provided to
```
s/is provided/provided/

stuartmorgan commented on shell/platform/common/cpp/public/flutter_texture_registrar.h:
```diff
@@ -0,0 +1,62 @@
+typedef struct FlutterDesktopTextureRegistrar*
+    FlutterDesktopTextureRegistrarRef;
+
+// Constructs a pixel buffer object for the plugin side, providing
```
I'm confused by a struct comment starting with a verb; there's no action associated with a struct.

stuartmorgan commented on shell/platform/common/cpp/public/flutter_texture_registrar.h:
```diff
@@ -0,0 +1,62 @@
+  size_t width;
+  // Height of the pixel buffer.
+  size_t height;
+} PixelBuffer;
```
This name should have a FlutterDesktop prefix.

stuartmorgan commented on shell/platform/common/cpp/public/flutter_texture_registrar.h:
```diff
@@ -0,0 +1,62 @@
+
+// The pixel buffer copy callback definition is provided to
+// the Flutter engine to copy the texture.
+typedef const PixelBuffer* (*FlutterTextureCallback)(size_t width,
```
FlutterDesktopTextureCallback

stuartmorgan commented on shell/platform/windows/BUILD.gn:
```diff
@@ -71,6 +73,7 @@ source_set("flutter_windows_source") {
 
   deps = [
     ":flutter_windows_headers",
+    "//flutter/fml",
```
This still needs to be addressed.

stuartmorgan commented on shell/platform/windows/external_texture_gl.cc:
```diff
@@ -0,0 +1,131 @@
+    FML_DCHECK(glDeleteTextures);
+    FML_DCHECK(glBindTexture);
+    FML_DCHECK(glTexParameteri);
+    FML_DCHECK(glTexImage2D);
```
As does this.

stuartmorgan commented:
Sorry, I completely missed that this had been updated and was waiting for another review. I've done some restructuring to the Windows APIs to get them closer to the final structure, which created conflicts here. If you can update this to the latest master and then ping me here, I'll make sure to review it ASAP.

jnschulze commented:
@stuartmorgan 
I‘ve just pushed an update.

stuartmorgan commented on shell/platform/common/cpp/client_wrapper/testing/stub_flutter_api.h:
```diff
@@ -67,6 +67,19 @@ class StubFlutterApi {
                                     void* user_data) {}
+
+  // Called for FlutterDesktopRegisterExternalTexture.
+  virtual int64_t RegisterExternalTexture(
```
TextureRegistrarRegisterExternalTexture (see comment about the function namen in the C header). Same for the ones below.

stuartmorgan commented on shell/platform/windows/window_state.h:
```diff
@@ -37,13 +38,24 @@ struct FlutterDesktopPluginRegistrar {
 };
 
+// State associated with the texture registrar.
+struct FlutterDesktopTextureRegistrar {
```
The use of structs for internal implementations is a legacy of the initial C++ embedding being a single file without much complexity; it shouldn't be perpetuated (and in fact is being slowly eliminated). New objects should be classes, and follow the normal class process (be declared in their own file, not expose ivars, contain relevant logic such as what's currently inlined in OnAcquireExternalTexture and the flutter_windows.cc functions, etc.).

stuartmorgan commented on shell/platform/windows/external_texture_gl.h:
```diff
@@ -0,0 +1,52 @@
+
+typedef struct ExternalTextureGLState ExternalTextureGLState;
+
+// An adaptation class of flutter engine and external texture interface.
```
This description isn't really clear to me, especially given that it doesn't implement an interface. Can you adjust this comment to more clearly describe this object's role in the texture system?

stuartmorgan commented on shell/platform/windows/external_texture_gl.h:
```diff
@@ -0,0 +1,52 @@
+
+  virtual ~ExternalTextureGL();
+
+  /**
```
Please use the same comment style as the rest of the platform/windows/ code

stuartmorgan commented on shell/platform/windows/external_texture_gl.cc:
```diff
@@ -0,0 +1,126 @@
+
+#include "flutter/shell/platform/windows/external_texture_gl.h"
+
+// OpenGL ES and EGL includes
```
This is obvious from the includes themselves; includes are generally only commented when there is something non-obvious that needs to be explained.

stuartmorgan commented on shell/platform/windows/external_texture_gl.cc:
```diff
@@ -0,0 +1,126 @@
+                                 GLenum type,
+                                 const void* data);
+
+struct GlProcs {
```
Needs a comment.

stuartmorgan commented on shell/platform/windows/external_texture_gl.cc:
```diff
@@ -0,0 +1,126 @@
+
+namespace flutter {
+
+typedef void (*glGenTexturesProc)(GLsizei n, GLuint* textures);
```
There should be an anonymous namespace her containing everything that doesn't correspond to a declaration in external_texture_gl.h

stuartmorgan commented on shell/platform/windows/external_texture_gl.cc:
```diff
@@ -0,0 +1,126 @@
+
+struct ExternalTextureGLState {
+  GLuint gl_texture = 0;
+  static const GlProcs procs;
```
I think I'm missing something here. What is the point of a static const reference in the struct declaration when this is all private to this file anyway? Why not just reference the functions directly?

jnschulze commented:
@stuartmorgan 
I've reworked this according to your comments. Please have another look. Thanks!

stuartmorgan commented on shell/platform/common/cpp/client_wrapper/include/flutter/plugin_registrar.h:
```diff
@@ -41,6 +42,10 @@ class PluginRegistrar {
 
+  // Returns the texture registrar to use for the plugin to render a pixel
+  // buffer.
+  TextureRegistrar* textures() { return texture_registrar_.get(); }
```
`texture_registrar()`

stuartmorgan commented on shell/platform/common/cpp/client_wrapper/include/flutter/texture_registrar.h:
```diff
@@ -0,0 +1,44 @@
+
+namespace flutter {
+
+// An interface used as an image source by texture widgets
```
Nit: Missing a '.' at the end of the sentence. This is the case for many comments in this file (and maybe the rest of the PR; please audit all the new comments for this).

stuartmorgan commented on shell/platform/common/cpp/client_wrapper/plugin_registrar_unittests.cc:
```diff
@@ -179,4 +236,37 @@ TEST(PluginRegistrarTest, ManagerRemovesOnDestruction) {
+  texture = test_api->GetFakeTexture(texture_id);
+  EXPECT_EQ(texture, nullptr);
+  EXPECT_EQ(test_api->textures_size(), static_cast<size_t>(0));
+}
```
This is a test of the texture registrar; it should be in a texture_registrar_unittests.cc file. The test here should just be that the getter on PluginRegistrar works.

stuartmorgan replied:
Also, there should be more than a single test. For instance, anything with error return values should have coverage where feasible (e.g., if unregister is going to return a bool, then there should be a test that unregistering a non-existent texture returns false as expected).

stuartmorgan commented on shell/platform/common/cpp/public/flutter_texture_registrar.h:
```diff
@@ -0,0 +1,63 @@
+
+// Registers a new texture with the Flutter engine and returns the texture ID,
+// The engine will use the |texture_callback|
+// function to copy the pixel buffer from the plugin caller.
```
s/plugin caller/caller/

stuartmorgan commented on shell/platform/common/cpp/public/flutter_texture_registrar.h:
```diff
@@ -0,0 +1,63 @@
+    void* user_data);
+
+// Unregisters an existing texture from the Flutter engine for a |texture_id|.
+// Returns true on success, false on failure.
```
What does failure mean here? It's not clear to me what a caller would do with this return value.

stuartmorgan commented on shell/platform/common/cpp/public/flutter_texture_registrar.h:
```diff
@@ -0,0 +1,63 @@
+    int64_t texture_id);
+
+// Marks that a new texture frame is available for a given |texture_id|.
+// Returns true on success, false on failure.
```
Same question here.

stuartmorgan commented on shell/platform/windows/external_texture_gl.cc:
```diff
@@ -0,0 +1,123 @@
+  bool valid;
+};
+
+static const GlProcs kGlProcs = [] {
```
This violates https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables

You can make a function with a static variable instead.

stuartmorgan commented on shell/platform/windows/external_texture_gl.cc:
```diff
@@ -0,0 +1,123 @@
+
+static const GlProcs kGlProcs = [] {
+  GlProcs p = {
+      .glGenTextures = reinterpret_cast<glGenTexturesProc>(
```
Labelled aggregate initialization is non-standard until C++20; the engine is still C++17.

stuartmorgan commented on shell/platform/windows/external_texture_gl.cc:
```diff
@@ -0,0 +1,123 @@
+      texture_callback_(width, height, user_data_);
+
+  if (!kGlProcs.valid || !pixel_buffer || !pixel_buffer->buffer) {
+    std::cerr << "Failed to copy pixel buffer from plugin." << std::endl;
```
s/plugin/provider/

stuartmorgan commented on shell/platform/windows/external_texture_gl.cc:
```diff
@@ -0,0 +1,123 @@
+  opengl_texture->name = state_->gl_texture;
+  opengl_texture->format = GL_RGBA8;
+  opengl_texture->destruction_callback = (VoidCallback) nullptr;
+  opengl_texture->user_data = static_cast<void*>(this);
```
Where/how is this used?

stuartmorgan commented on shell/platform/windows/flutter_windows_texture_registrar.cc:
```diff
@@ -0,0 +1,63 @@
+
+bool FlutterWindowsTextureRegistrar::UnregisterTexture(int64_t texture_id) {
+  auto it = textures_.find(texture_id);
+
```
Nit: The C++ style guide prefers using horizontal whitespace only to separate conceptually distinct code; this function is only really doing one thing, so doesn't need the whitespace (and certainly a find and a check of the find result should not be separated). Same with PopulateTexture below.

stuartmorgan commented on shell/platform/windows/flutter_windows_texture_registrar.h:
```diff
@@ -0,0 +1,52 @@
+
+class FlutterWindowsEngine;
+
+class FlutterWindowsTextureRegistrar {
```
Needs a declaration comment.

stuartmorgan commented on shell/platform/windows/flutter_windows_texture_registrar.h:
```diff
@@ -0,0 +1,52 @@
+ private:
+  FlutterWindowsEngine* engine_ = nullptr;
+
+  // A map keeping track of all registered textures.
```
`// All registered textures, keyed by their IDs.`

The fact that it's a map is trivially obvious from the code; the meaning of the int64_t is not.

stuartmorgan commented on shell/platform/windows/window_state.h:
```diff
@@ -12,8 +12,16 @@
 }  // namespace flutter
 
+// Opaque reference to the internal texture registrar
+struct FlutterDesktopTextureRegistrar {
```
There's no need for the struct at all; you can just use a pointer to the class directly as the ref. See the engine ref for an example of that.

The goal is to eliminate all of these structs.

stuartmorgan commented on shell/platform/windows/window_state.h:
```diff
@@ -37,6 +45,9 @@ struct FlutterDesktopPluginRegistrar {
   flutter::FlutterWindowsEngine* engine = nullptr;
 
+  // The plugin texture registrar handle given to API clients.
+  std::unique_ptr<FlutterDesktopTextureRegistrar> texture_registrar;
```
Then this isn't needed, and you can just get the registrar instance from |engine| when you need to return a ref.

stuartmorgan commented on shell/platform/windows/flutter_windows.cc:
```diff
@@ -255,3 +255,28 @@ void FlutterDesktopMessengerSetCallback(FlutterDesktopMessengerRef messenger,
   messenger->dispatcher->SetMessageCallback(channel, callback, user_data);
 }
+
+FlutterDesktopTextureRegistrarRef FlutterDesktopRegistrarGetTextureRegistrar(
```
I would expect there to also be a `FlutterDesktopEngineGetTextureRegistrar`, and corresponding accessor in the C++ wrapper.

stuartmorgan commented on shell/platform/windows/external_texture_gl.cc:
```diff
@@ -0,0 +1,134 @@
+  auto glTexParameteri = reinterpret_cast<glTexParameteriProc>(
+      eglGetProcAddress("glTexParameteri"));
+  auto glTexImage2D =
+      reinterpret_cast<glTexImage2DProc>(eglGetProcAddress("glTexImage2D"));
```
You're doing all of these lookups every time the function is called, which defeats the point of having the struct below be `static`.

The basic structure of this function should be something like:
```
static bool sInitialized = false;
static struct GlProcs sProcs = {};
if (!sInitialized) {
  [ do all the lookup to populate sProcs ]
  sInitialized = true;
}
return sProcs;
```

jnschulze replied:
I agree. But a test showed that the compiler appears to be smart enough to not call `eglGetProcAddress` again " (as it's const here).
But it might to easier to just change it as opposed to finding evidence ;)

stuartmorgan replied:
I'm curious how you did that test, because I don't see how the compiler could be "smart enough" to optimize away calls to a function that could have arbitrary side effects...

jnschulze replied:
I created a "spy" function logging all invocations to `eglGetProcAddress`.

stuartmorgan replied:
I'm not sure what that means exactly. (In fact, the possibility of the function doing logging is an illustration of exactly why I would not expect the optimization you are describing to be possible.)

stuartmorgan commented on shell/platform/windows/external_texture_gl.cc:
```diff
@@ -0,0 +1,134 @@
+  opengl_texture->target = GL_TEXTURE_2D;
+  opengl_texture->name = state_->gl_texture;
+  opengl_texture->format = GL_RGBA8;
+  opengl_texture->destruction_callback = (VoidCallback) nullptr;
```
C-style casts are forbidden by the C++ style guide; this should use the appropriate C++ cast. (Does it actually need a cast? IIRC `nullptr` can be assigned to a function pointer directly).

stuartmorgan commented on shell/platform/windows/external_texture_gl.cc:
```diff
@@ -0,0 +1,134 @@
+  opengl_texture->name = state_->gl_texture;
+  opengl_texture->format = GL_RGBA8;
+  opengl_texture->destruction_callback = (VoidCallback) nullptr;
+  opengl_texture->user_data = static_cast<void*>(this);
```
I asked previous where/how this was used; you marked that question resolved but didn't answer it (either there or with a comment in the code).

jnschulze replied:
There's a comment above the highlighted block ("Populate the texture object used by the engine."). In fact, it's used by `MakeSkSurfaceFromBackingStore` (`shell/platform/embedder/embedder.cc`).

stuartmorgan replied:
My question is about the line it's attached to. Engine `user_data` void* fields are for the use of the caller, allowing arbitrary data to be resurfaced for later use (e.g., in callbacks). Nothing in this file ever casts `user_data` back to an `ExternalTextureGL` which makes me think it's unused.

An in fact, looking at `MakeSkSurfaceFromBackingStore` it appears that this is a context object for the callback that you've set to `nullptr` on the line above.

stuartmorgan commented on shell/platform/common/cpp/public/flutter_texture_registrar.h:
```diff
@@ -0,0 +1,63 @@
+#endif
+
+// Opaque reference to a texture registrar.
+typedef struct FlutterDesktopTextureRegistrar*
```
Add a
```
struct FlutterDesktopTextureRegistrar;
```

line above this, so the struct is declared. (This is needed to avoid warnings in some compilation modes when nothing else declares the struct later, as is the case for Windows in this iteration since no such struct ever exists.)

stuartmorgan commented on shell/platform/windows/flutter_windows_texture_registrar.h:
```diff
@@ -0,0 +1,53 @@
+class FlutterWindowsEngine;
+
+// An object managing the registration of an external texture.
+class FlutterWindowsTextureRegistrar {
```
This class needs unit tests; sorry I missed that last time.

jnschulze commented:
Hi @stuartmorgan, I've addressed the requested changes.

stuartmorgan commented on shell/platform/windows/external_texture_gl.cc:
```diff
@@ -93,8 +93,8 @@ bool ExternalTextureGL::PopulateTexture(size_t width,
   opengl_texture->format = GL_RGBA8;
-  opengl_texture->destruction_callback = (VoidCallback) nullptr;
-  opengl_texture->user_data = static_cast<void*>(this);
+  opengl_texture->destruction_callback = [](void* user_data) {};
```
Why not just nullptr?

jnschulze replied:
I wasn’t sure if there are null checks in `embedder.cc` and Skia’s `SkImage::MakeFromTexture` but I just checked and apparently this can be just set to nullptr.

stuartmorgan commented on shell/platform/windows/flutter_windows_texture_registrar_unittests.cc:
```diff
@@ -0,0 +1,46 @@
+// TODO Mock FlutterEngineRegisterExternalTexture etc.
+// TEST(FlutterWindowsTextureRegistrarTest, RegisterTexture) {
+//   EXPECT_TRUE(true);
+// }
```
We don't generally land commented out code. It's fine to include a TODO for more tests explaining what they should cover and what's needed to implement them, but the commented out stub of a test function isn't really useful.

chinmaygarde commented:
While the Embedder API usage in this patch looks fine, Stuart and I were having a conversation about the suitability of this API as a general purpose mechanism for providing textures to the Flutter Engine. Towards this end, this is going to be prohibitively slow and memory intensive. This is because each texture update in this API requires a host (main memory) to device (VRAM) memory copy of the texture contents (via `glTexImage2D`). This in itself is pretty slow and memory intensive for interactive applications. However, most existing plugins (camera, video, etc..) take textures from sources where the texture data is already in device memory.  Such plugins will now have to copy that texture data from device to host memory to specify it via this API which will in turn transfer it back to device memory. This round-trip is slow and memory intensive and serves no purpose.

A general purpose texture API would allow for the specification of OpenGL texture handles directly. These handles can be reference counted (see `FlutterOpenGLTexture::destruction_callback` for ideas). This way, textures specified to the Flutter Engine incur no copies at the API level. This is how Android is wired up. Notice that the downside of this approach is that plugins need to be aware of OpenGL specific implementation details. So, if in the future, this platform supports multiple rendering backends (like Direct3D), the plugins will need to figure out how to provide D3D textures via this API. iOS works around this by dealing with texture information in the form of CoreVideo pixel buffers. These pixel buffers can be cast to either an OpenGL texture or a Metal texture (Android still has this, as yet, unsolved problem).

My initial LGTM was based on the review of the usage of the Flutter Embedder API, and it’s fine. But, I recommend devising an API that does not require unnecessary round trips.

This API very obviously covers some use-cases. Specifically the case where textures are specified once upfront but then these textures are not updated frequently (or at all). I suspect for use cases like the camera or video plugins, this will take a ton of memory and time to render.

jnschulze commented:
@chinmaygarde 
I absolutely agree.
However, some time ago, I stumbled across a comment of yours (“The pixel buffer interface is intentional and allows plugins that provide external textures for inline composition to provide these textures in a manner that is agnostic to client rendering API used by Flutter.”, see https://github.com/flutter/flutter/issues/59410#issuecomment-643860848) which made me think that rendering API agnosticism is quite a concern and plugin implementers should just deal with copying pixel buffers.

But I’m looking forward to having a zero-copy solution as well.

stuartmorgan commented:
The issue is that for Windows it's not clear that there is a rendering-agnostic way to do performant texture handoff. If this version is going to be so slow that it can't actually support common use cases for Texture, then we should figure out what the performant, non-agnostic API looks like so that we can see if, for instance, there's a good way to support both. It's not ideal to have to do that (it means some plugins would need to do rewrites for an eventual DirectX transition) but it's better than not being able to support those use cases at all.

I'm reluctant to land this if we don't think it's a viable approach and don't already have a plan in place to extend it to a performant version. So I think two questions I'd like to resolve before landing it are:
- Do we know of real-world use cases that this version will adequately support?
- What does the performant, non-agnostic API look like in detail, so that we can see how it aligns with this version. If it could be added later, great, but if it would require breaking changes in this one, then I would like to do that up front and avoid the churn for clients.

jnschulze commented:
> The issue is that for Windows it's not clear that there is a rendering-agnostic way to do performant texture handoff. If this version is going to be so slow that it can't actually support common use cases for Texture, then we should figure out what the performant, non-agnostic API looks like so that we can see if, for instance, there's a good way to support both. It's not ideal to have to do that (it means some plugins would need to do rewrites for an eventual DirectX transition) but it's better than not being able to support those use cases at all.
> 
> I'm reluctant to land this if we don't think it's a viable approach and don't already have a plan in place to extend it to a performant version. So I think two questions I'd like to resolve before landing it are:
> 
> * Do we know of real-world use cases that this version will adequately support?
> * What does the performant, non-agnostic API look like in detail, so that we can see how it aligns with this version. If it could be added later, great, but if it would require breaking changes in this one, then I would like to do that up front and avoid the churn for clients.

I've been creating a cross-platform video playback plugin that's using ffmpeg for decoding.
While it appears to be possible to implement a zero-copy approach somehow (see https://stackoverflow.com/questions/57242800/how-to-convert-an-ffmpeg-texture-to-open-gl-texture-without-copying-to-cpu-memor) there's no universal approach for ffmpeg (considering all available decoding backends etc.) that entirely avoids copying pixel buffers. Therefore I decided to live with copying the AVFrame contents to Flutter for now.

For 1920 * 1080 * 32 color bits * 60 fps this means roughly 475 Megabytes per second which might be a lot on some low-clocked ARM platform but on a today's DDR4 desktop machine this is just nothing.

Anyhow, thanks for you input. I will reconsider the plugin-side API.

stuartmorgan commented:
Interesting; good to know that it's viable (even if inefficient) for at least some video use cases.

That does suggest that if there's a reasonable API design that would allow for clients to choose either the slow-agnostic or fast-openGL approach, that would be good to support.

chinmaygarde commented:
@stuartmorgan From PR triage, it is not clear if this should land. I am adding the WIP tag for now but feel free to update as necessary.

stuartmorgan commented:
I'll look again next week to see if it seems like there's a simple way (maybe with some renaming to make the pixel-based APIs less generically named, to leave room for future additions) that we could move forward with this close to as-is. Since it sounds like it has potential long-term value for certain use cases, as long as we do it in a way that won't cause problems later we could move forward with this.

peerwaya commented:
@stuartmorgan did you get a chance to see if there is a simple way to move forward with this?

stuartmorgan commented:
The Windows alpha ended up monopolizing my time. I haven't forgotten this though; it's high on my list to look at next week.

stuartmorgan commented on shell/platform/common/cpp/public/flutter_texture_registrar.h:
```diff
@@ -0,0 +1,64 @@
+// function to copy the pixel buffer from the caller.
+FLUTTER_EXPORT int64_t FlutterDesktopTextureRegistrarRegisterExternalTexture(
+    FlutterDesktopTextureRegistrarRef texture_registrar,
+    FlutterDesktopTextureCallback texture_callback,
```
At this layer, I think you could reasonably make this extensible to other provider types by replacing this with a discriminated union (see `FlutterRendererConfig` for instance). That way this first version could be a pixel buffer texture type, but we could add other types with other callback signatures in the future.

stuartmorgan commented on shell/platform/common/cpp/client_wrapper/include/flutter/texture_registrar.h:
```diff
@@ -0,0 +1,44 @@
+namespace flutter {
+
+// An interface used as an image source by texture widgets.
+class Texture {
```
For the wrapper, this could become a `PixelBufferTexture`

stuartmorgan commented on shell/platform/common/cpp/client_wrapper/include/flutter/texture_registrar.h:
```diff
@@ -0,0 +1,44 @@
+  virtual ~TextureRegistrar() {}
+
+  // Registers a |texture| object and returns the ID for that texture.
+  virtual int64_t RegisterTexture(Texture* texture) = 0;
```
And then this could take a `std::variant` of texture types, typedef'd to a set list of known texture types. As long as we only added new types to the end of it, that shouldn't break most code when updated (we could explicitly document that code should expect more types to be added, so that people don't do things like write exhaustive visitors of the texture variant—not that I can think of why someone would want to do that anyway).

I don't think there's a good way to do it via a class hierarchy given that the signature of the equivalent to `CopyPixelBuffer` isn't likely to be the same.

stuartmorgan reviewed (changes requested):
@jnschulze Sorry for the long delay; I finally got back to this. Comments below for a small set of changes that I think would give us a workable API to extend this in the future without breaking changes. Let me know if anything doesn't make sense, or if it isn't practical because of something I didn't take into account in implementation.

jnschulze commented:
@stuartmorgan 
This sounds like a clean solution. I will provide an update by end of this week.

peerwaya commented:
@jnschulze trust you are doing fine. Did you get a chance to look into this over the weekend?

fanchou commented:
Any updates?

jnschulze commented:
@fanchou @peerwaya 
No, unfortunately no updates yet. I'll take care of it this weekend.

jnschulze commented:
@stuartmorgan 
It's been a while, but I've finally had some time to update this.

Besides renaming the Texture class to PixelBufferTexture etc., I've added an additional change that ensures `MarkTextureFrameAvailable` gets always called on the platform thread by posting a task to respective task runner (given that it must be called from that thread).

Please have a look.

leielyq commented:
Come on

vjeson commented:
> @bigdig thanks for your feedback.
> I haven‘t updated my demo plugin yet to see if it‘s working as expected.
> But what about the `TestPixelBufferTexture` in `texture_registrar_unittests.cc`?

I can't get it to work, could update your example demo? And it's better to how to use it in multithreading with a video player

vjeson commented:
> @bigdig thanks for your feedback.
> I haven‘t updated my demo plugin yet to see if it‘s working as expected.
> But what about the `TestPixelBufferTexture` in `texture_registrar_unittests.cc`?

I can't get it to work, could update your example demo? And it's better to how to use it in multithreading with a video player

stuartmorgan commented:
This is still on my radar, but I've been mostly OOO for the last few weeks, and will continue to be through the end of the year. This is a substantial PR that I need a significant block of time to re-review, which unfortunately makes it hard to fit in when my time is so limited. It may be the beginning of January before I'll be able to get back to this, but if I'm able to do it sooner I definitely will.

jnschulze commented:
@vjeson 
> I can't get it to work, could you update your example demo? And it's better to how to use it in multithreading with a video player

I've just updated the demo and added a simple timer that periodically calls `MarkTextureFrameAvailable` to demonstrate that it can be called from arbitrary threads. See https://github.com/jnschulze/flutter-playground/blob/ae3f0965dc1e4ab3a266fcbad382e319ac8efd0d/windows_texture_test/windows/windows_texture_test_plugin.cpp#L159

stuartmorgan commented on shell/platform/common/cpp/client_wrapper/include/flutter/texture_registrar.h:
```diff
@@ -0,0 +1,58 @@
+#define FLUTTER_SHELL_PLATFORM_COMMON_CPP_CLIENT_WRAPPER_INCLUDE_FLUTTER_TEXTURE_REGISTRAR_H_
+
+#include <flutter_texture_registrar.h>
+#include <stdint.h>
```
Nit: prefer `cstdint`. We've been moving toward standardizing on C++ versions of includes.

stuartmorgan commented on shell/platform/common/cpp/client_wrapper/include/flutter/texture_registrar.h:
```diff
@@ -0,0 +1,58 @@
+  // Returns a FlutterDesktopPixelBuffer that contains the actual pixel data.
+  // The intended surface size is specified by |width| and |height|.
+  virtual const FlutterDesktopPixelBuffer* CopyPixelBuffer(size_t width,
+                                                           size_t height) = 0;
```
Apologies if this is rehashing earlier discussion that I've forgotten and can't find, but is there any expectation that we'll need other methods on this delegate? If not, it seems like it would be simpler to have PixelBufferTexture just take a `std::function` to call to populate it, rather than making an interface for the one function.

stuartmorgan commented on shell/platform/common/cpp/client_wrapper/include/flutter/texture_registrar.h:
```diff
@@ -0,0 +1,58 @@
+  std::unique_ptr<PixelBufferTextureDelegate> delegate_;
+};
+
+typedef std::variant<PixelBufferTexture> TextureVariant;
```
Please add a comment here noting that other texture types may be added in the future, so code should not assume that the current list of supported types are the only possible types.

stuartmorgan commented on shell/platform/common/cpp/client_wrapper/texture_registrar_impl.h:
```diff
@@ -0,0 +1,35 @@
+  TextureRegistrarImpl(TextureRegistrarImpl const&) = delete;
+  TextureRegistrarImpl& operator=(TextureRegistrarImpl const&) = delete;
+
+  int64_t RegisterTexture(TextureVariant* texture) override;
```
Flutter engine style is to put a comment on overridden methods that just gives the class that the method is declared it. (In this case there's currently only one, but we try to be consistent about it for convenience of readers and for futureproofing.)

stuartmorgan commented on shell/platform/common/cpp/public/flutter_texture_registrar.h:
```diff
@@ -0,0 +1,81 @@
+} FlutterDesktopPixelBuffer;
+
+// The pixel buffer copy callback definition provided to
+// the Flutter engine to copy the texture.
```
Please document the meaning of the parameters and return here.

stuartmorgan commented on shell/platform/common/cpp/public/flutter_texture_registrar.h:
```diff
@@ -0,0 +1,81 @@
+typedef struct {
+  FlutterDesktopTextureType type;
+  union {
+    FlutterDesktopPixelBufferTextureConfig pixel_buffer;
```
Nit: Let's call this `pixel_buffer_config` so that it doesn't sound like a `FlutterDesktopPixelBuffer` instance.

stuartmorgan commented on shell/platform/common/cpp/public/flutter_texture_registrar.h:
```diff
@@ -0,0 +1,81 @@
+typedef struct FlutterDesktopTextureRegistrar*
+    FlutterDesktopTextureRegistrarRef;
+
+typedef enum {
```
Please add a comment describing the purpose of the enum. It wouldn't hurt to explicitly mention that other types may be added in the future, to make in clear why there's a one-element enum.

stuartmorgan commented on shell/platform/embedder/embedder.h:
```diff
@@ -2119,6 +2121,7 @@ typedef struct {
   FlutterEngineTraceEventDurationBeginFnPtr TraceEventDurationBegin;
   FlutterEngineTraceEventDurationEndFnPtr TraceEventDurationEnd;
   FlutterEngineTraceEventInstantFnPtr TraceEventInstant;
+  FlutterEnginePostPlatformThreadTaskFnPtr PostPlatformThreadTask;
```
New entries in embedder.h structs must always go at the end for binary compatibility reasons.

stuartmorgan replied:
There should be a function version of this API as well; this struct is intended as a 1:1 alternative to the functions above.

stuartmorgan commented on shell/platform/embedder/embedder.cc:
```diff
@@ -1691,6 +1691,28 @@ void FlutterEngineTraceEventInstant(const char* name) {
   fml::tracing::TraceEventInstant0("flutter", name);
 }
 
+FlutterEngineResult FlutterEnginePostPlatformThreadTask(
```
Why would the embedder need to ask the engine to post a task to the platform thread rather than doing so directly?

jnschulze replied:
So I guess it’s safe to directly post a task onto the Windows platform task runner (with target time set to `fml::TimePoint::Now()`)?

jnschulze replied:
I'm stuck here. Can you give me a hint how to achieve that?
For the time being, I've fixed `shell/platform/embedder/embedder.h` according to your comments.

stuartmorgan replied:
> Can you give me a hint how to achieve that?

Achieve what? The issue is that I'm not clear on what the purpose of this API is. I.e. I don't understand the use case for doing this:

> I guess it’s safe to directly post a task onto the Windows platform task runner [...]?

Why would an embedder want to use the engine's task runner for platform thread work?

jnschulze replied:
Thanks for your quick reply. This is basically needed as the `MarkTextureFrameAvailable` is expected to be called from the platform thread (see check at
https://github.com/flutter/engine/blob/b304148d465e9d30e0543128248f0f033ada4fec/shell/common/shell.cc#L953). So I was looking for a solution to post a task onto the platform task runner.
The most obvious way was to add the `PostPlatformThreadTask` method.

iskakaushik replied:
If the only call site for this method is https://github.com/flutter/engine/pull/19405/files#diff-e0363cfd79295a13710333f2e74c37d754b2477bf8312306bcc53cffdbdd0ab8R56, couldn't you do `engine_->task_runner()` and post this task there? Given that this is the same platform task runner (and you could avoid adding an extra api).

jnschulze replied:
But the task runner https://github.com/flutter/engine/blob/52901f23b4b40367efb36768e4a39a624b97f955/shell/platform/windows/task_runner_win32.h#L39 expects a `FlutterTask` (declared here: https://github.com/flutter/engine/blob/52901f23b4b40367efb36768e4a39a624b97f955/shell/platform/embedder/embedder.h#L763) as opposed to a closure?

stuartmorgan replied:
> This is basically needed as the `MarkTextureFrameAvailable` is expected to be called from the platform thread [...]. So I was looking for a solution to post a task onto the platform task runner.

This is what I was trying to ask in the first place though: why do you want to use the platform task runner—a Flutter engine concept—to do things on the platform thread? Why not just use platform APIs directly?

I'm assuming the use case here is a plugin; plugins know the platform thread (it's the thread they were created on, and from which they must call ~all Flutter APIs). If they are running code on other threads, then they are already using platform-specific threading APIs, so why not use them to handle the cross-thread posting?

(Contrast with the method for posting to the renderer thread, which exists because the embedder may not create that thread, and thus not have any way to access it.)

stuartmorgan replied:
Oh, I missed that this was being used internally. I'll apply my question above to that code then, and ask: why is the embedding handling the threading, instead of the caller?

jnschulze replied:
Because as opposed to iOS/macOS platforms, we don't have someting like GCD's `dispatch_async(dispatch_get_main_queue(), ...)` on Windows, so plugin implementers have no API for "getting back" to the main thread.

So the Client Wrapper should either expose methods for explicitly posting tasks to the platform thread or methods like `MakeTextureFrameAvailable` need to be made thread agnostic as a short-term solution.
I can't think of any scenario where the exposed Texture API would give us any benefit over no API unless the rendering-related functions can be called from "anywhere".

jnschulze replied:
Look at the [flutter-webrtc](https://github.com/flutter-webrtc/flutter-webrtc) project for example.
Their macOS video renderer implemention at https://github.com/flutter-webrtc/flutter-webrtc/blob/33277024c46327bbb04fe43bec5d2866b99097f4/common/darwin/Classes/FlutterRTCVideoRenderer.m#L210 indeed uses dispatch_async() for invoking `[registry textureFrameAvailable]` on the main thread.

stuartmorgan replied:
> so plugin implementers have no API for "getting back" to the main thread.

Sorry, my macOS background is ingrained enough that it's hard for me to remember that there's no concept of a generally available runloop at all. I was thinking the issue was finding the thread, and forgetting that there's no way to actually run code on it without the runloop that's owned by the runner driving that.

I think the main question then is whether to plumb this through embedder.h, or pipe it in to our task runner as @iskakaushik suggested above (adding new API there as necessary). @chinmaygarde Thoughts on the embedder.h version?

stuartmorgan replied:
Per chat, Chinmay prefers not adding the embedder.h API (since the render thread version is actually legacy), so we should plumb it into the Windows embedding's task runner instead.

Also, you should clearly document at both the C and the C++ wrapper levels that those specific methods can be called from any thread, but also that it's the caller's responsibility to avoid races with other Flutter API calls they make. Notable ones to call out would be:
- Unregistering a texture while it's possible that a background thread is still using it.
- Shutting down the engine while it's possible that a background thread is producing texture frames.

jnschulze replied:
I've just extended the Windows TaskRunner to allow posting callbacks directly (see https://github.com/flutter/engine/pull/19405/commits/6a8a20e6e5bf139f004dbcdc7d59a0f3dbdcae87) and refactored the texture registrar accordingly besides reverting my changes to `embedder.{h,cc}`.

stuartmorgan reviewed (changes requested):
This is looking really good; sorry for the long delay. I think the only major question is about the new platform task API on embedder.h, everything else should be pretty minor.

stuartmorgan reviewed (changes requested):
Looking really good. I think all the fundamentals are resolved now, and it's just final cleanup.

clarkezone commented:
Based on a quick scan through this looks like it should work with the forthcoming UWP embedder.  FYI the PR to introduce the bulk of the UWP embedder impl is here https://github.com/flutter/engine/pull/23573 if you have a chance to take a look for potential issues / breaks would much appreciate it.

stuartmorgan reviewed (approved):
LGTM! Thanks for the contribution, and especially for your patience!

jnschulze commented:
@stuartmorgan 
thanks for your reviewing effort!
Please consider that there is a corresponding PR for copying the newly introduced `flutter_texture_registrar.h` (https://github.com/flutter/flutter/pull/61098). May I ask you to review it as well and ensure that it will be merged at the right time?

stuartmorgan commented:
> Please consider that there is a corresponding PR for copying the newly introduced `flutter_texture_registrar.h` ([flutter/flutter#61098](https://github.com/flutter/flutter/pull/61098)). May I ask you to review it as well and ensure that it will be merged at the right time?

I forgot until I was reviewing that PR that we actually only copy the files on the list on Windows (rather than a whole directory, as with most of the other headers for desktop). That means that we need to be careful here to avoid breaking Windows builds since we can't tie that PR to the engine roll that will happen automatically at some later point; Flutter master between the engine roll and that PR landing would be unable to build Windows apps because there would be an include of a missing header. And we actually need to update the [recipe](https://flutter.googlesource.com/recipes/+/refs/heads/master/recipes/engine.py#1245) as well.

Could you make a new PR that just adds a new, empty (except for license boilerplate) flutter_texture_registrar.h to the Windows build, and ping me for a review? That way we can do this as:
- Land that empty header
- Update the recipe (I'll take care of that once the empty header is in place)
- Land the `flutter` tool PR to copy the header that nothing is using yet
- Land this

jnschulze commented:
@stuartmorgan done, see https://github.com/flutter/engine/pull/23623

stuartmorgan commented:
Recipe change out for review: https://flutter-review.googlesource.com/c/recipes/+/9680

escamoteur commented:
We are currently working on an FFI wrapper for the WebGPU API as a more flexible approach to OpenGL and till I stumbled upon this issue I wasn't aware that its currently not possbile to use a Widget with external Textturebuffer on Windows.
Is this already avalable on MacOs and Linux?

stuartmorgan commented:
> Is this already avalable on MacOs and Linux?

https://github.com/flutter/flutter/issues/64188

vjeson commented:
> @vjeson
> 
> > I can't get it to work, could you update your example demo? And it's better to how to use it in multithreading with a video player
> 
> I've just updated the demo and added a simple timer that periodically calls `MarkTextureFrameAvailable` to demonstrate that it can be called from arbitrary threads. See https://github.com/jnschulze/flutter-playground/blob/52005d38e6935c4bbafe07b7f6f5346ca7cc30de/windows_texture_test/windows/windows_texture_test_plugin.cpp#L166

Thanks for your demo,I still have another issue, I'm migrating flutter-webrtc for windows, it uses event_sink onFrame event.
void FlutterVideoRenderer::OnFrame(scoped_refptr<RTCVideoFrame> frame) {
  if (!first_frame_rendered) {
    if (event_sink_) {
      EncodableMap params;
      params[EncodableValue("event")] = "didFirstFrameRendered";
      params[EncodableValue("id")] = texture_id_;
      //event_sink_->Success(EncodableValue(params));
    }
Here I got a thread exception, we need to switch to UI thread to run event_sink_->Success .
Just as the doc official demo, I don't known how to call it on desktop embedding,It seems we didn't export such method for desktop embedding.
https://flutter.dev/docs/development/platform-integration/platform-channels#channels-and-platform-threading
Maybe we need to have a method like this ? task_runners_.GetPlatformTaskRunner()->RunsTasksOnCurrentThread()

jnschulze commented:
> https://flutter.dev/docs/development/platform-integration/platform-channels#channels-and-platform-threading
> Maybe we need to have a method like this ? task_runners_.GetPlatformTaskRunner()->RunsTasksOnCurrentThread()

You're right. Unfortunately, the client wrapper doesn't expose a method like `PostTask`.

While it should be fairly simple to extend the client wrapper accordingly, we shouldn't discuss it here as it's a general Windows platform issue.

escamoteur commented:
> > Is this already avalable on MacOs and Linux?
> 
> [flutter/flutter#64188](https://github.com/flutter/flutter/issues/64188)

So there isn't even an issue for MacOs?

jnschulze commented:
> > > Is this already avalable on MacOs and Linux?
> > 
> > 
> > [flutter/flutter#64188](https://github.com/flutter/flutter/issues/64188)
> 
> So there isn't even an issue for MacOs?

Only Windows and Linux are missing. macOS support was introduced by https://github.com/flutter/engine/pull/8507

stuartmorgan commented:
The handling of the new header has landed end-to-end, so this is good to land as soon as the tree re-opens 🎉

dahabit commented:
@stuartmorgan Thank you for your support, I really appreciate it, this issue was blocking using for making cool Windows desktop apps with Flutter.

can't wait to see it land in the stable channel

nmbus commented:
What would it require for this to gain some zero-copy interop with an OpenGL texture/FBO (and/or other APIs) ? Adapted the provided example to have a texture backed by an OpenGL renderer and calls to glReadPixels are prohibitively expensive for any real time rendering (around 10ms for a 1920x1080 texture). Pixel buffers aren't suited for such use.

stuartmorgan commented:
> What would it require for this to gain some zero-copy interop with an OpenGL texture/FBO (and/or other APIs) ?

A design doc with one or more concrete proposals for possible texture sources, including exploring what that API would look like, how it would work with Angle, and what the impact on that API of an eventual migration to a DirectX Skia backend would be.
